### PR TITLE
`Theme`: Use intersection observer to render a placeholder when not visible

### PR DESCRIPTION
**Pull request by xavier-lc**
Related to https://github.com/Automattic/wp-calypso/issues/95248

## Proposed Changes

* Update the `theme` component so it's lazy loaded when it's visible or within the next viewport height.

## Why are these changes being made?

* The `/themes` page loads ~34MB of theme screenshots on page load, because it's rendering 100 themes for each page. This is too much, and lazy loading the component allows us to reduce that to ~5MB.

* The initial [PR](https://github.com/Automattic/wp-calypso/pull/95881) had to be [reverted](https://github.com/Automattic/wp-calypso/pull/95951) because it was breaking an E2E test, which I fixed on https://github.com/Automattic/wp-calypso/pull/95984/commits/562d8b9afb871a3c535e0a469dc53850e8942c4a.

* I also noticed that the lazy loading was not working properly on the logged in theme showcase, with only the visible themes being loaded. This happened because the intersection observer's default root element is the document viewport, and since the logged in theme showcase has its own scroll, the intersection is not properly calculated. In order to fix it, I provided the theme showcase wrapper ref through a react context.

## Testing Instructions

* On the `Files changed` tab, enable the `Hide whitespace` option so the diff is more readable

* Open the live preview while logged out
* Open the dev tools, go to the network tab, filter by `.png`
* Go to `/themes`
* Check that only 8 theme screenshots are loaded (could be more, depending on the screen size), resulting in way less image data transferred (`5.1MB`) than what's happening on prod (`33.6MB`):
* Scroll down, check that the theme images are loaded 
* Check that it works as well for mobile devices
* Repeat the previous steps on the logged in theme showcase

|Before|After|
|---|---|
|Untitled|Untitled|

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by BogdanUngureanu** [comment]
Thanks for the change.

Looks like there's still a regression on the Global View logged-in:
- Go to https://container-practical-cray.calypso.live/themes (Logged-in)
- Use a mobile resolution (e.g. iPhone 12)
- Refresh the page
- Scroll a bit
- Notice that the images are not loading.

**Comment by xavier-lc** [comment]
@BogdanUngureanu nice catch! On mobile devices, the showcase wrapper scrolls out of view when you go down the page, so there's no intersection. I've fixed it so it uses the document viewport as root on mobile devices.

**Comment by BogdanUngureanu** [comment]
Looks like there are still issues on mobile :(

https://github.com/user-attachments/assets/5f6d4bf5-33f1-417e-89d6-176ae5ffb367

**Comment by xavier-lc** [comment]
@BogdanUngureanu good catch! I've added a new [breakpoint util](https://github.com/Automattic/wp-calypso/pull/95984/commits/b1b9b09b1afca312cc49fdcf836bdc0549f9b942) because the layout rules apply both for mobile width or height.

**Comment by BogdanUngureanu** [comment]
Thanks, @xavier-lc 

I did a new session and I've noticed two more issues :(
- We seem to make a new /plans request on each theme card display (on mobile)
- I'm seeing more theme thumbnail image flashes on mobile resolutions - every theme image is flashing on loading, giving a non-optimal scrolling UI.

https://github.com/user-attachments/assets/1530cee5-4e5b-4605-bd71-af9920b02d26

**Comment by dsas** [comment]
Not blocking, just checking. It looks like this works on the logged-out theme showcase, and it works on the logged in global theme showcase (/themes) but it has no effect on logged-in site specific theme showcase (/themes/:siteSlug).

Happy for that to be a follow-up PR, but thought it was worth mentioning in case it's not what is expected.

**Comment by xavier-lc** [comment]
> * We seem to make a new /plans request on each theme card display (on mobile) - this is more serious since it's happening on Logged out too (returning a lot of 403 error responses)

Addressed on https://github.com/Automattic/wp-calypso/pull/95984/commits/f9aebd88881124c108c67dfbbfa619343f35ae6e, preventing `usePlans` from being fired with every new set of themes.

File: packages/data-stores/src/plans/queries/use-plans.ts [hunk 1]
```diff
@@ -73,6 +73,7 @@ function usePlans( {
 				} )
 			);
 		},
+		staleTime: 3600000, // 1 hour in ms
```
Comment (mmtr) [inline packages/data-stores/src/plans/queries/use-plans.ts:76]: Why is this changed needed? Does this mean that if we ship some API changes to update the pricing or the name of some plans, they will not be applied for up to 1 hour? Because if that's the case, I think we need to discuss this change more deeply with other teams. With the current code, the expectation is that API changes to the plans will be reflected immediately in Calypso.

**Review by mmtr — COMMENTED** [review]
This tests well for me, but I'm bit concerned about the changes of the `usePlans` hook.

Also, as @dsas noted, theme screenshots are not lazy loaded in the site-specific TS (`/themes/:site`). I think that's expected and will be addressed in a follow-up, but has not been confirmed.

File: packages/data-stores/src/plans/queries/use-plans.ts (hunk 1 above)
Comment (xavier-lc) [reply packages/data-stores/src/plans/queries/use-plans.ts:76]: Bogdan mentioned that there were refetches when lazy loading a new batch of themes:

> We seem to make a new /plans request on each theme card display (on mobile) - this is more serious since it's happening on Logged out too (returning a lot of 403 error responses)

After some investigation I founds that it happens because the `ThemeTierUpgradeBadge` triggers a `usePlans` call when it's rendered, so that's what I set a `staleTime` of one hour.

> Does this mean that if we ship some API changes to update the pricing or the name of some plans, they will not be applied for up to 1 hour?

No, they would be applied immediately, but users that are already navigating the site without actually reloading it would not see the new data until one hour has gone by after the last hook call.

Anyway, the value can be set to one minute as well, I don't think a request per minute while the user scrolls the themes is something to be concerned. Wdyt?

**Comment by xavier-lc** [comment]
> This tests well for me, but I'm bit concerned about the changes of the `usePlans` hook.

Sure, I answered that on the comment.

> Also, as @dsas noted, theme screenshots are not lazy loaded in the site-specific TS (`/themes/:site`). I think that's expected and will be addressed in a follow-up, but has not been confirmed.

Yeah, I didn't look into that because the PR had enough changes without it, I'll tackle it on a followup PR.

File: packages/data-stores/src/plans/queries/use-plans.ts (hunk 1 above)
Comment (mmtr) [reply packages/data-stores/src/plans/queries/use-plans.ts:76]: > After some investigation I founds that it happens because the `ThemeTierUpgradeBadge` triggers a `usePlans` call when it's rendered, so that's what I set a `staleTime` of one hour.

This made me think 🤔 It seems that we're only interested in lazy loading the screenshots, right? So can we then move the lazy loading logic from the whole theme card to only the screenshots? That way the `ThemeTierUpgradeBadge` component (and the `usePlans` hook) won't be affected by these changes. 

I also wonder, why these changes increase the number of requests that `ThemeTierUpgradeBadge` component make to `/plans` from one per page load to one per each theme card? That might be a symptom of an underlying issue that we should fix (instead of masking it with a `staleTime`).

Comment (xavier-lc) [reply packages/data-stores/src/plans/queries/use-plans.ts:76]: > So can we then move the lazy loading logic from the whole theme card to only the screenshots? That way the `ThemeTierUpgradeBadge` component (and the `usePlans` hook) won't be affected by these changes.

I thank that would still mean replacing the whole content rendered by `ThemeCard` with what's on the `renderPlaceholder` function of the `Theme` component. So the `badge` would be still re-rendered and triggering the `usePlans` hook.



> I also wonder, why these changes increase the number of requests that `ThemeTierUpgradeBadge` component make to `/plans` from one per page load to one per each theme card?

As I understand it, it's the expected `useQuery` behavior. Having the default `stleTime: 0` means that react query will always trigger a background request to the plans endpoint when a new instance of the `useQuery` hook is rendered, since the cached data is considered stale. That doesn't happen currently on production because all the themes are rendered at the same time and so at no point there's a new `usePlans` instance.

**Comment by mmtr** [comment]
Closing this in favor of https://github.com/Automattic/wp-calypso/pull/96961

File: packages/data-stores/src/plans/queries/use-plans.ts (hunk 1 above)
Comment (mmtr) [reply packages/data-stores/src/plans/queries/use-plans.ts:76]: > I thank that would still mean replacing the whole content rendered by `ThemeCard` with what's on the `renderPlaceholder` function of the Theme component.

I meant the specific component used for the screenshot returned by the `renderScreenshot` function. But that made wonder: why are over-complicating this with an `InView` component when we already have a native solution with the `loading="lazy"` attribute that makes all the work? So I opened an alternative solution: https://github.com/Automattic/wp-calypso/pull/96961
